### PR TITLE
Add machine learning benchmarks

### PR DESCRIPTION
**Pull request by mariaschuld**
Adds the last app benchmark.

At some stage we should add tests, but I checked that training works in all three frameworks (i.e., the loss gets smaller and the parameters change, and the objects are the correct kind of tensors).

### CODE REVIEW

File: benchmarks/asv/app_suite.py [hunk 1]
```diff
@@ -134,3 +136,73 @@ def peakmem_minvertex_heavy(self):
         hyperparams = {'n_layers': self.n_layers,
                        'graph': self.graph}
         benchmark_qaoa(hyperparams)
+
+
+class ML_light:
+    """Benchmark a hybrid quantum-classical machine learning application with a small dataset."""
+
+    n_features = 4
+    n_samples = 20
```
Comment (mariaschuld) [inline benchmarks/asv/app_suite.py:147]: takes a minute or so for each function below.

File: benchmarks/asv/app_suite.py [hunk 2]
```diff
@@ -134,3 +136,73 @@ def peakmem_minvertex_heavy(self):
...
+        hyperparams = {'n_layers': self.n_features,
+                       'n_samples': self.n_samples}
+        benchmark_machine_learning(hyperparams)
+
+
+class ML_heavy:
+    """Benchmark a hybrid quantum-classical machine learning application with a large dataset."""
+
+    n_features = 10
+    n_samples = 100
```
Comment (mariaschuld) [inline benchmarks/asv/app_suite.py:171]: Takes about 5 min per function below.

File: benchmarks/benchmark_functions/machine_learning.py [hunk 3]
```diff
@@ -17,69 +17,142 @@
...
 import pennylane as qml
-import tensorflow as tf
-import torch
 from pennylane import numpy as pnp
-from .default_settings import _core_defaults
+from .default_settings import _ml_defaults
 
 
-qml.enable_tape()
+def _machine_learning_autograd(quantum_model, data):
```
Comment (mariaschuld) [inline benchmarks/benchmark_functions/machine_learning.py:24]: The three functions below basically implement gradient descent by hand.

File: benchmarks/asv/app_suite.py [hunk 4]
```diff
@@ -134,3 +136,73 @@ def peakmem_minvertex_heavy(self):
         hyperparams = {'n_layers': self.n_layers,
                        'graph': self.graph}
         benchmark_qaoa(hyperparams)
+
+
+class ML_light:
```
Comment (albi3ro) [inline benchmarks/asv/app_suite.py:141]: I'm wanting to change the word "light" to "small" so selecting "light" doesn't also select "lightning.qubit".

File: benchmarks/asv/app_suite.py [hunk 5]
```diff
@@ -134,3 +136,73 @@ def peakmem_minvertex_heavy(self):
...
+    """Benchmark a hybrid quantum-classical machine learning application with a small dataset."""
+
+    n_features = 4
+    n_samples = 20
+
+    def time_ml_autograd_light(self):
```
Comment (albi3ro) [inline benchmarks/asv/app_suite.py]: Is there a reason we separate the different interfaces into different benchmarks?

It'd be nice to easily see how the different interfaces compare to each other in speed on the same plot.

File: benchmarks/benchmark_functions/default_settings.py [hunk 6]
```diff
@@ -129,3 +130,30 @@ def _qaoa_defaults(hyperparams):
...
+	diff_method = hyperparams.pop('diff_method', 'best')
+	device = hyperparams.pop('device', 'default.qubit')
+
+	# if device name is given, create device
+	if isinstance(device, str):
+		device = qml.device(device, wires=n_features)
+
+	# data
+	x0 = np.array([[np.random.normal(loc=-1, scale=1) for i in range(n_features)] for j in range(n_samples // 2)])
+	x1 = np.array([[np.random.normal(loc=1, scale=1) for i in range(n_features)] for j in range(n_samples // 2)])
```
Comment (albi3ro) [inline benchmarks/benchmark_functions/default_settings.py]: why not 
```
x = np.random.normal(loc=-1, scale=1, size = (n_features, n_samples // 2, 2) )
```

Seems like that would be a more efficient way of creating the same data.

File: benchmarks/asv/app_suite.py (hunk 2 above)
Comment (albi3ro) [reply benchmarks/asv/app_suite.py:171]: For the heavier functions, we should specify [benchmarking specifications](https://asv.readthedocs.io/en/stable/benchmarks.html#id5) like `repeat`, `number`, or `sample_time`.  We should run as may iterations of the heavy functions.

File: benchmarks/benchmark_functions/machine_learning.py [hunk 7]
```diff
@@ -17,69 +17,142 @@
...
-	return NotImplemented
+	w_quantum = pnp.array(random(size=(n_features, n_features)), requires_grad=True)
+	w_classical = pnp.array(random(size=(n_features, n_features)), requires_grad=True)
 
+	gradient_fn_wq = qml.grad(average_loss, argnum=0)
+	gradient_fn_wc = qml.grad(average_loss, argnum=1)
 
-def _machine_learning_torch(circuit, params):
-	return NotImplemented
+	for i in range(50):
```
Comment (albi3ro) [inline benchmarks/benchmark_functions/machine_learning.py]:
```suggestion
	for _ in range(50):
```

Looks like `i` isn't getting used.

File: benchmarks/benchmark_functions/machine_learning.py [hunk 8]
```diff
@@ -17,69 +17,142 @@
...
+		w_quantum = w_quantum - 0.05 * gradient_fn_wq(w_quantum, w_classical)
+		w_classical = w_classical - 0.05 * gradient_fn_wc(w_quantum, w_classical)
+
+
+def _machine_learning_tf(quantum_model, data):
+	"""ML example with tensorflow interface."""
+
+	import tensorflow as tf
```
Comment (albi3ro) [inline benchmarks/benchmark_functions/machine_learning.py:54]: I'm assuming if we import here, then we can run the other functions without importing tensorflow?  

Can we change `gradient.py` to import this way, so we can run autograd benchmarking without having to get tensorflow and torch installed?

File: benchmarks/benchmark_functions/machine_learning.py [hunk 9]
```diff
@@ -17,69 +17,142 @@
...
+			prediction = hybrid_model(x, w_quantum, w_classical)
+			c = c + (prediction - y) ** 2
+		return c / len(data)
+
+	n_features = len(data[0][0])
+
+	w_quantum = tf.Variable(random(size=(n_features, n_features)), dtype=tf.double)
+	w_classical = tf.Variable(random(size=(n_features, n_features)), dtype=tf.double)
+
+	for i in range(50):
```
Comment (albi3ro) [inline benchmarks/benchmark_functions/machine_learning.py]:
```suggestion
	for _ in range(50):
```

File: benchmarks/benchmark_functions/machine_learning.py [hunk 10]
```diff
@@ -17,69 +17,142 @@
...
+			prediction = hybrid_model(x, w_quantum, w_classical)
+			c += (prediction - y) ** 2
+		return c / len(data)
+
+	n_features = len(data[0][0])
+
+	w_quantum = torch.tensor(random(size=(n_features, n_features)), requires_grad=True, dtype=torch.double)
+	w_classical = torch.tensor(random(size=(n_features, n_features)), requires_grad=True, dtype=torch.double)
+
+	for i in range(50):
```
Comment (albi3ro) [inline benchmarks/benchmark_functions/machine_learning.py]:
```suggestion
	for _ in range(50):
```

File: benchmarks/benchmark_functions/machine_learning.py [hunk 11]
```diff
@@ -17,69 +17,142 @@
...
+		* 'device': device on which the circuit is run, or valid device name. Defaults to 'default.qubit.
 
-		* 'measurement': measurement function like `qml.expval(qml.PauliZ(0)))`
+		* 'interface': name of the interface to use. Defaults to 'autograd'.
 
 	num_repeats (int): How often the same circuit is evaluated in a for loop. Default is 1.
 	"""
 
-	device, diff_method, interface, params, template, measurement = _core_defaults(hyperparams)
+	data, device, diff_method, interface = _ml_defaults(hyperparams)
```
Comment (albi3ro) [inline benchmarks/benchmark_functions/machine_learning.py:141]: Should we move setting defaults to a `setup` function, so we don't time that?  It's probably trivial compared to the rest of the evaluation, but it might help focus the benchmark on what we most care about.

**Review by albi3ro — COMMENTED** [review]
Thank you!   

I left a few comments, but they are mostly optional suggestions.

File: benchmarks/asv/app_suite.py (hunk 4 above)
Comment (mariaschuld) [reply benchmarks/asv/app_suite.py:141]: What about `lightweight`? That is longer, but unique :)

File: benchmarks/asv/app_suite.py (hunk 5 above)
Comment (mariaschuld) [reply benchmarks/asv/app_suite.py]: Good point. Let me do that!

File: benchmarks/benchmark_functions/default_settings.py (hunk 6 above)
Comment (mariaschuld) [reply benchmarks/benchmark_functions/default_settings.py]: Of course :)

File: benchmarks/benchmark_functions/machine_learning.py (hunk 8 above)
Comment (mariaschuld) [reply benchmarks/benchmark_functions/machine_learning.py:54]: Done.

**Comment by mariaschuld** [comment]
Thanks for the helpful comments, which I implemented. I am just not sure what I must do for this one:

 > specify benchmarking specifications like repeat, number, or sample_time

File: benchmarks/benchmark_functions/machine_learning.py (hunk 11 above)
Comment (mariaschuld) [reply benchmarks/benchmark_functions/machine_learning.py:141]: Good point...the idea was that someone should be able to run the functions without asv though. Can we use `setup` but still maintain compatibility with other ways to run?

File: benchmarks/asv/app_suite.py (hunk 4 above)
Comment (mariaschuld) [reply benchmarks/asv/app_suite.py:141]: Or `lightapp` which allows us to only run the light app bms...

Comment (soranjh) [reply benchmarks/asv/app_suite.py:141]: I think we have `_light` almost everywhere in the function/class names. We can then select "_light" for the tests and be fine without using long names or names that are not as descriptive as "light".

File: benchmarks/asv/app_suite.py [hunk 12]
```diff
@@ -134,3 +136,50 @@ def peakmem_minvertex_heavy(self):
...
+    n_samples = 100
+
+    def time_ml_heavy(self, interface):
+        """Time 50 training steps of a hybrid quantum machine learning example."""
+        hyperparams = {'n_layers': self.n_features,
+                       'n_samples': self.n_samples,
+                       'interface': interface}
+        benchmark_machine_learning(hyperparams)
+
+    def peakmem_ml_training_heavy(self, interface):
```
Comment (soranjh) [inline benchmarks/asv/app_suite.py]: Maybe remove `_training` in the name for consistency with other function names.

File: benchmarks/benchmark_functions/default_settings.py [hunk 13]
```diff
@@ -129,3 +130,30 @@ def _qaoa_defaults(hyperparams):
...
+
+
+def _ml_defaults(hyperparams):
+	"""Uses hyperparameters or defaults to construct the components of the machine learning benchmark.
+
+	Args:
+		hyperparams (dict): hyperparameters provided by user
+	"""
+	# get hyperparameters or set default values
+	n_features = hyperparams.pop('n_wires', 4)
```
Comment (soranjh) [inline benchmarks/benchmark_functions/default_settings.py]: 'n_wires' should be 'n_features'?

File: benchmarks/asv/app_suite.py (hunk 12 above)
Comment (mariaschuld) [inline benchmarks/asv/app_suite.py]:
```suggestion
    def peakmem_ml_heavy(self, interface):
```

File: benchmarks/benchmark_functions/default_settings.py (hunk 13 above)
Comment (mariaschuld) [reply benchmarks/benchmark_functions/default_settings.py]: Thanks, true!

Comment (mariaschuld) [inline benchmarks/benchmark_functions/default_settings.py]:
```suggestion
	n_features = hyperparams.pop('n_features', 4)
```

File: benchmarks/asv/app_suite.py (hunk 2 above)
Comment (mariaschuld) [reply benchmarks/asv/app_suite.py:171]: Shall we do this in a separate PR to be consistent?

File: benchmarks/asv/app_suite.py (hunk 4 above)
Comment (albi3ro) [reply benchmarks/asv/app_suite.py:141]: Good idea on using `_light`.  I'll go ahead and use that.

File: benchmarks/asv/app_suite.py (hunk 2 above)
Comment (albi3ro) [reply benchmarks/asv/app_suite.py:171]: Sounds good.

File: benchmarks/benchmark_functions/machine_learning.py (hunk 11 above)
Comment (albi3ro) [reply benchmarks/benchmark_functions/machine_learning.py:141]: Even if you look at `timeit` or other benchmarking tools, they often support a "setup" component.  And if the particular tool does not have a "setup" component, we can just run both functions.

**Review by albi3ro — APPROVED** [review]
Looks good to me :)
